### PR TITLE
 server: always create a liveness record before starting up

### PR DESCRIPTION
**Pull request by irfansharif**
Previously it used to be the case that it was possible for a node to be
up and running, and for there to be no corresponding liveness record for
it. This was a very transient situation as liveness records are created
for a given node as soon as it out its first heartbeat. Still, given
that this could take a few seconds, it lent to a lot of complexity in
our handling of node liveness where we had to always anticipate the
possibility of there being no corresponding liveness record for a given
node (and thus creating it if necessary).

It was originally thought we'd only be able to guarantee persisting a
liveness record for a joining node if we had infrastructure like the
join rpc, introduced in #52526, but staring at it a bit harder we can
simply ensure we heartbeat our liveness record at least once before
fully spinning up our server (aka opening the RPC floodgates).

Having a liveness record for each node always present is a crucial
building block for long running migrations (#48843). There the intention
is to have the orchestrator process look towards the list of liveness
records for an authoritative view of cluster membership. Previously when
it was possible for an active member of the cluster to not have a
corresponding liveness record (no matter how unlikely or short-lived in
practice), we could not generate such a view.

Release justification: low risk, high benefit changes to existing functionality
Release note: None

---

First commit is from #53713, both are necessary to remove the hack in the decommissioning roachtest.

### CODE REVIEW

**Comment by irfansharif** [comment]
@tbg, I'm a bit confused. I'm not sure why why this wasn't possible before. The only thing I can think of to make us not want to to this is the following scenario:

- n1-n3 running v21.1, initialized and running
- n4, running v21.1 joins the cluster, gets a node ID, persists it to disk
- n4 crashes before creating its liveness record
- operator is able to roll n1-n3 to v21.2 and finalize upgrade because n4's liveness record is not present
- n4 is never decommissioned, operator never needed to (also it wasn't really possible, even before this PR, seeing as how n4 crashed before being able to create a liveness record)
- n4 is started back up, still running v21.1, finds the cluster/node id on disk and now happily creates its liveness record/continues on merrily.
- ...boom?

While shitty, this is also a bug that already exists today. I've actually seen it happen in https://github.com/cockroachlabs/support/issues/584 where a node was quickly added+removed before it was able to persist a liveness record for itself. So that said, is there another reason why we shouldn't do what we're doing in this PR?

**Comment by irfansharif** [comment]
I'm not sure why I'm seeing gossip/locality-address fail with my changes. I suspect it's due to https://github.com/cockroachdb/cockroach/issues/42741, but I could be wrong.

File: pkg/server/server.go [hunk 1]
```diff
@@ -1538,6 +1536,32 @@ func (s *Server) Start(ctx context.Context) error {
 		}
 	})
 
+	// Begin the node liveness heartbeat. Add a callback which records the
+	// local store "last up" timestamp for every store whenever the liveness
+	// record is updated. We're sure to do this before we open RPC
+	// floodgates.
```
Comment (tbg) [inline pkg/server/server.go:1542]: Not guaranteed since it's async, so add that it's best effort

File: pkg/server/server.go [hunk 2]
```diff
@@ -1538,6 +1536,32 @@ func (s *Server) Start(ctx context.Context) error {
...
+	var livenessOnce sync.Once
```
Comment (tbg) [inline pkg/server/server.go:1543]: nit: don't leak `livenessOnce` into main scope, do

```go
livenessRecordCreated := make(chan struct{}, 1)
{
  // all the other stuff
}
if initialStart { ... }
```

**Comment by tbg** [comment]
This PR is good and there's no good reason we haven't done this yet. But as you've found out, it also doesn't give us the invariant that we need, which is that the liveness record exists *before* the node can ever consider itself part of the cluster.

I don't really understand the problem you're seeing regarding the "sending itself heartbeats". Until we're in modeOperational, the node can not receive replicas, so it shouldn't ever be consulted for anything. Gossip should be active by this point so the KV client should be able to route to the correct nodes. Can you post some more details?

> Is there some grpc magic "don't send an external rpc if headed for 127.0.0.1" that's coming into play here?

Yes there is! See https://github.com/cockroachdb/cockroach/blob/a3284b792b0631c1572dcd50c29f346d7b984664/pkg/rpc/context.go#L481

But if we rely on that for things working, there's a problem.

Ah, I understand now - the issue is in the case in which this node is the one bootstrapping the local cluster. In that case, in the new code, we're trying to synchronously heartbeat the liveness record but the gRPC server does not yet allow RPCs. Ok, makes sense. Add a comment to that effect and simply move the liveness record stuff past the `modeOperational` line. Since we are only doing best effort here, it makes little difference (except of course that it works!)

**Comment by irfansharif** [comment]
Abandoning this PR in favor of #53842.

**Pull request closed without merging**